### PR TITLE
STR-503: use `OsRng` consistently

### PR DESCRIPTION
**Pull request by AaronFeickert**
## Description

Switches all uses of `thread_rng()` to `OsRng`. Standardizes existing uses of `OsRng`.

### Type of Change

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [ ] Refactor

## Checklist

-   [x] I have performed a self-review of my code.
-   [x] I have commented my code where necessary.
-   [x] I have updated the documentation if needed.
-   [x] My changes do not introduce new warnings.
-   [ ] I have added tests that prove my changes are effective or that my feature works.
-   [x] New and existing tests pass with my changes.

## Related Issues

None.

### CODE REVIEW

File: bin/strata-cli/src/recovery.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ use bdk_wallet::{
     miniscript::{descriptor::DescriptorKeyParseError, Descriptor},
     template::DescriptorTemplateOut,
 };
-use rand::{thread_rng, Rng};
+use rand::{rngs::OsRng, RngCore};
```
Comment (storopoli) [inline bin/strata-cli/src/recovery.rs:16]: Why do we need `RngCore` here?

File: crates/bridge-tx-builder/src/operations.rs [hunk 2]
```diff
@@ -281,7 +281,7 @@ mod tests {
             "should work if the number of scripts is not an exact power of 2"
         );
 
-        let secret_key = SecretKey::new(&mut rand::thread_rng());
+        let secret_key = SecretKey::new(&mut rand::rngs::OsRng);
```
Comment (storopoli) [inline crates/bridge-tx-builder/src/operations.rs:284]: We should move this to an import
```suggestion
        let secret_key = SecretKey::new(&mut OsRng);
```

File: bin/strata-client/src/extractor.rs [hunk 3]
```diff
@@ -619,7 +621,7 @@ mod tests {
             "some deposits should have been randomly dispatched"
         );
 
-        let needle_index = rand::thread_rng().gen_range(0..dispatched_deposits.len());
+        let needle_index = rand::rngs::OsRng.gen_range(0..dispatched_deposits.len());
```
Comment (storopoli) [inline bin/strata-client/src/extractor.rs:624]: We should move this to an import
```suggestion
        let needle_index = OsRng.gen_range(0..dispatched_deposits.len());
```

File: bin/strata-client/src/extractor.rs [hunk 4]
```diff
@@ -575,7 +577,7 @@ mod tests {
             break;
         }
 
-        let mut rng = rand::thread_rng();
+        let mut rng = rand::rngs::OsRng;
```
Comment (storopoli) [inline bin/strata-client/src/extractor.rs:580]:
```suggestion
        let mut rng = OsRng;
```

Comment (storopoli) [reply bin/strata-client/src/extractor.rs:580]: Also do we need this line at all? Can't we just pass `OsRng` to whatever was taking `rng` here?

File: crates/btcio/src/writer/builder.rs [hunk 5]
```diff
@@ -402,7 +402,7 @@ pub fn generate_key_pair(
     secp256k1: &Secp256k1<secp256k1::All>,
 ) -> Result<UntweakedKeypair, anyhow::Error> {
     let mut rand_bytes = [0; 32];
-    rand::thread_rng().fill_bytes(&mut rand_bytes);
+    rand::rngs::OsRng.fill_bytes(&mut rand_bytes);
```
Comment (storopoli) [inline crates/btcio/src/writer/builder.rs:405]:
```suggestion
    OsRng.fill_bytes(&mut rand_bytes);
```

File: crates/btcio/src/writer/builder.rs [hunk 6]
```diff
@@ -465,7 +465,7 @@ fn sign_reveal_transaction(
     )?;
 
     let mut randbytes = [0; 32];
-    rand::thread_rng().fill_bytes(&mut randbytes);
+    rand::rngs::OsRng.fill_bytes(&mut randbytes);
```
Comment (storopoli) [inline crates/btcio/src/writer/builder.rs:468]:
```suggestion
    OsRng.fill_bytes(&mut randbytes);
```

File: crates/crypto/src/lib.rs [hunk 7]
```diff
@@ -45,7 +45,7 @@ mod tests {
     #[test]
     fn test_schnorr_signature_pass() {
         let secp = Secp256k1::new();
-        let mut rng = rand::thread_rng();
+        let mut rng = rand::rngs::OsRng;
```
Comment (storopoli) [inline crates/crypto/src/lib.rs:48]:
```suggestion
        let mut rng = OsRng;
```
same as above, do we need the `let mut rng`?

File: crates/evmexec/src/el_payload.rs [hunk 8]
```diff
@@ -125,7 +125,7 @@ mod tests {
     #[test]
     fn into() {
         let mut rand_data = vec![0u8; 1024];
-        rand::thread_rng().fill_bytes(&mut rand_data);
+        rand::rngs::OsRng.fill_bytes(&mut rand_data);
```
Comment (storopoli) [inline crates/evmexec/src/el_payload.rs:128]:
```suggestion
        OsRng.fill_bytes(&mut rand_data);
```

File: crates/evmexec/src/engine.rs [hunk 9]
```diff
@@ -419,7 +419,7 @@ mod tests {
     }
 
     fn random_execution_payload_v1() -> ExecutionPayloadV1 {
-        let mut rng = rand::thread_rng();
+        let mut rng = rand::rngs::OsRng;
```
Comment (storopoli) [inline crates/evmexec/src/engine.rs:422]:
```suggestion
        let mut rng = OsRng;
```
same as above, do we need the `let mut rng`?

File: crates/primitives/src/bridge.rs [hunk 10]
```diff
@@ -162,7 +162,7 @@ impl BorshDeserialize for Musig2PartialSig {
 
 impl<'a> Arbitrary<'a> for Musig2PartialSig {
     fn arbitrary(_u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
-        let secret_key = SecretKey::new(&mut rand::thread_rng());
+        let secret_key = SecretKey::new(&mut rand::rngs::OsRng);
```
Comment (storopoli) [inline crates/primitives/src/bridge.rs:165]:
```suggestion
        let secret_key = SecretKey::new(&mut OsRng);
```

File: bin/strata-client/src/extractor.rs (hunk 4 above)
Comment (storopoli) [reply bin/strata-client/src/extractor.rs:580]: Maybe we can do `(&mut OsRng)`? Yeah that would be annoying also, big ugly and hard-to-visually-parse thing.

File: crates/primitives/src/hash.rs [hunk 11]
```diff
@@ -47,7 +47,7 @@ mod tests {
 
     #[test]
     fn test_sha256d_equivalence() {
-        let mut rng = rand::thread_rng();
+        let mut rng = rand::rngs::OsRng;
```
Comment (storopoli) [inline crates/primitives/src/hash.rs:50]:
```suggestion
        let mut rng = OsRng;
```

File: crates/primitives/src/l1.rs [hunk 12]
```diff
@@ -839,7 +839,7 @@ impl<'a> Arbitrary<'a> for TaprootSpendPath {
                 };
 
                 // Generate a random secret key and derive the internal key
-                let secret_key = SecretKey::new(&mut rand::thread_rng());
+                let secret_key = SecretKey::new(&mut rand::rngs::OsRng);
```
Comment (storopoli) [inline crates/primitives/src/l1.rs:842]:
```suggestion
                let secret_key = SecretKey::new(&mut OsRng);
```

File: crates/primitives/src/l1.rs [hunk 13]
```diff
@@ -915,7 +915,7 @@ mod tests {
 
         let num_possible_networks = possible_networks.len();
 
-        let (secret_key, _) = SECP256K1.generate_keypair(&mut rand::thread_rng());
+        let (secret_key, _) = SECP256K1.generate_keypair(&mut rand::rngs::OsRng);
```
Comment (storopoli) [inline crates/primitives/src/l1.rs:918]:
```suggestion
        let (secret_key, _) = SECP256K1.generate_keypair(&mut OsRng);
```

File: crates/primitives/src/l1.rs [hunk 14]
```diff
@@ -930,7 +930,7 @@ mod tests {
             let invalid_network = match network {
                 Network::Bitcoin => {
                     // get one of the testnets
-                    let index = rand::thread_rng().gen_range(1..num_possible_networks);
+                    let index = rand::rngs::OsRng.gen_range(1..num_possible_networks);
```
Comment (storopoli) [inline crates/primitives/src/l1.rs:933]:
```suggestion
                    let index = OsRng.gen_range(1..num_possible_networks);
```

File: crates/primitives/src/l1.rs [hunk 15]
```diff
@@ -1189,7 +1189,7 @@ mod tests {
         let output_key_parity = Parity::Even;
 
         // Generate a random internal key
-        let secret_key = SecretKey::new(&mut rand::thread_rng());
+        let secret_key = SecretKey::new(&mut rand::rngs::OsRng);
```
Comment (storopoli) [inline crates/primitives/src/l1.rs:1192]:
```suggestion
        let secret_key = SecretKey::new(&mut OsRng);
```

File: crates/primitives/src/relay/util.rs [hunk 16]
```diff
@@ -198,7 +198,7 @@ mod tests {
         let txid: BitcoinTxid = generator.generate();
         let scope = Scope::V0PubNonce(txid);
         let payload: Musig2PubNonce = generator.generate();
-        let keypair = Keypair::new(SECP256K1, &mut rand::thread_rng());
+        let keypair = Keypair::new(SECP256K1, &mut rand::rngs::OsRng);
```
Comment (storopoli) [inline crates/primitives/src/relay/util.rs:201]:
```suggestion
        let keypair = Keypair::new(SECP256K1, &mut OsRng);
```

File: crates/proof-impl/btc-blockspace/src/block.rs [hunk 17]
```diff
@@ -161,7 +161,7 @@ mod tests {
         // }
 
         let parsed_tx: ProtocolOperation = ArbitraryGenerator::new().generate();
-        let r = rand::thread_rng().gen_range(1..block.txdata.len()) as u32;
+        let r = rand::rngs::OsRng.gen_range(1..block.txdata.len()) as u32;
```
Comment (storopoli) [inline crates/proof-impl/btc-blockspace/src/block.rs:164]:
```suggestion
        let r = OsRng.gen_range(1..block.txdata.len()) as u32;
```

File: crates/proof-impl/btc-blockspace/src/merkle.rs [hunk 18]
```diff
@@ -73,7 +73,7 @@ mod tests {
 
     #[test]
     fn test_merkle_root() {
-        let mut rng = rand::thread_rng();
+        let mut rng = rand::rngs::OsRng;
```
Comment (storopoli) [inline crates/proof-impl/btc-blockspace/src/merkle.rs:76]:
```suggestion
        let mut rng = OsRng;
```

File: crates/rocksdb-store/src/l1/db.rs [hunk 19]
```diff
@@ -536,7 +536,7 @@ mod tests {
             l1_txs.push(l1_tx);
         }
 
-        let offset = rand::thread_rng().gen_range(0..total_num_blocks);
+        let offset = rand::rngs::OsRng.gen_range(0..total_num_blocks);
```
Comment (storopoli) [inline crates/rocksdb-store/src/l1/db.rs:539]:
```suggestion
        let offset = OsRng.gen_range(0..total_num_blocks);
```

File: crates/state/src/l1/header_verification.rs [hunk 20]
```diff
@@ -283,7 +283,7 @@ mod tests {
         // TODO: figure out why passing btc_params to `check_and_update_full` doesn't work
         let btc_params: BtcParams = MAINNET.clone().into();
         let h1 = get_difficulty_adjustment_height(1, chain.start, &btc_params);
-        let r1 = rand::thread_rng().gen_range(h1..chain.end);
+        let r1 = rand::rngs::OsRng.gen_range(h1..chain.end);
```
Comment (storopoli) [inline crates/state/src/l1/header_verification.rs:286]:
```suggestion
        let r1 = OsRng.gen_range(h1..chain.end);
```

File: crates/state/src/l1/header_verification.rs [hunk 21]
```diff
@@ -295,7 +295,7 @@ mod tests {
     #[test]
     fn test_get_difficulty_adjustment_height() {
         let start = 0;
-        let idx = rand::thread_rng().gen_range(1..1000);
+        let idx = rand::rngs::OsRng.gen_range(1..1000);
```
Comment (storopoli) [inline crates/state/src/l1/header_verification.rs:298]:
```suggestion
        let idx = OsRng.gen_range(1..1000);
```

File: crates/test-utils/src/bridge.rs [hunk 22]
```diff
@@ -31,7 +31,7 @@ pub fn generate_keypairs(count: usize) -> (Vec<PublicKey>, Vec<SecretKey>) {
     let mut pubkeys_set: HashSet<PublicKey> = HashSet::new();
 
     while pubkeys_set.len() != count {
-        let sk = SecretKey::new(&mut rand::thread_rng());
+        let sk = SecretKey::new(&mut rand::rngs::OsRng);
```
Comment (storopoli) [inline crates/test-utils/src/bridge.rs:34]:
```suggestion
        let sk = SecretKey::new(&mut OsRng);
```

File: crates/test-utils/src/bridge.rs [hunk 23]
```diff
@@ -198,7 +198,7 @@ pub fn generate_sec_nonce(
 /// This is used to generate random order for indices in a list (for example, list of pubkeys,
 /// nonces, etc.)
 pub fn permute<T: Clone>(list: &mut [T]) {
-    let num_permutations = rand::thread_rng().gen_range(0..list.len());
+    let num_permutations = rand::rngs::OsRng.gen_range(0..list.len());
```
Comment (storopoli) [inline crates/test-utils/src/bridge.rs:201]:
```suggestion
    let num_permutations = OsRng.gen_range(0..list.len());
```

File: crates/test-utils/src/lib.rs [hunk 24]
```diff
@@ -27,7 +27,7 @@ impl ArbitraryGenerator {
     }
 
     pub fn new_with_size(n: usize) -> Self {
-        let mut rng = rand::thread_rng();
+        let mut rng = rand::rngs::OsRng;
```
Comment (storopoli) [inline crates/test-utils/src/lib.rs:30]:
```suggestion
        let mut rng = OsRng;
```

File: crates/tx-parser/src/filter.rs [hunk 25]
```diff
@@ -180,7 +180,7 @@ mod test {
         // Create controlblock
         let secp256k1 = Secp256k1::new();
         let mut rand_bytes = [0; 32];
-        rand::thread_rng().fill_bytes(&mut rand_bytes);
+        rand::rngs::OsRng.fill_bytes(&mut rand_bytes);
```
Comment (storopoli) [inline crates/tx-parser/src/filter.rs:183]:
```suggestion
        OsRng.fill_bytes(&mut rand_bytes);
```

File: tests/cooperative-bridge-flow.rs [hunk 26]
```diff
@@ -82,7 +82,7 @@ async fn full_flow() {
     let unspent_utxos_prewithdrawal = user.agent().get_unspent_utxos().await;
     event!(Level::DEBUG, event = "got unspent utxos from requester before withdrawal", num_unspent_utxos = %unspent_utxos_prewithdrawal.len());
 
-    let assigned_operator_idx = rand::thread_rng().gen_range(0..num_operators) as OperatorIdx;
+    let assigned_operator_idx = rand::rngs::OsRng.gen_range(0..num_operators) as OperatorIdx;
```
Comment (storopoli) [inline tests/cooperative-bridge-flow.rs:85]:
```suggestion
    let assigned_operator_idx = OsRng.gen_range(0..num_operators) as OperatorIdx;
```

File: tests/cooperative-bridge-out-flow.rs [hunk 27]
```diff
@@ -47,7 +47,7 @@ async fn withdrawal_flow() {
 
     event!(Level::INFO, event = "created user to initiate withdrawal", user_x_only_pk = ?user_x_only_pk);
 
-    let assigned_operator_idx = rand::thread_rng().gen_range(0..num_operators) as OperatorIdx;
+    let assigned_operator_idx = rand::rngs::OsRng.gen_range(0..num_operators) as OperatorIdx;
```
Comment (storopoli) [inline tests/cooperative-bridge-out-flow.rs:50]:
```suggestion
    let assigned_operator_idx = OsRng.gen_range(0..num_operators) as OperatorIdx;
```

**Review by storopoli — CHANGES_REQUESTED** [review]
One big question that I could not find in the `OsRng` docs: since we are replacing `thread_rng` for `OsRng` in a bunch of stuff in multi-threaded context, **is `OsRng` thread-safe?**. Maybe my question might be dumb since that's what I would at most expect from an OS-based CSPRNG.

Also have minor nits, mostly related to the import flying under the radar since they are being fully-referenced instead of importing in the top. (I know that this is not your doing, but would be nice to fix these if possible)

File: bin/strata-cli/src/recovery.rs (hunk 1 above)
Comment (AaronFeickert) [reply bin/strata-cli/src/recovery.rs:16]: It's required to provide `fill_bytes`. I prefer this paradigm for setting random byte arrays over relying on specific `gen`-based distributions.

File: crates/bridge-tx-builder/src/operations.rs (hunk 2 above)
Comment (AaronFeickert) [reply crates/bridge-tx-builder/src/operations.rs:284]: Agreed. My initial goal was to minimize changes for easier review, but I agree that I prefer imports. Will update.

File: bin/strata-client/src/extractor.rs (hunk 4 above)
Comment (AaronFeickert) [reply bin/strata-client/src/extractor.rs:580]: Yeah, a separate assignment isn't particularly useful; the intent was just to minimize changes. I actually prefer `&mut OsRng` and will change.

**Comment by AaronFeickert** [comment]
I've never seen any issued raised with thread safety of `OsRng`, but don't have a definitive answer. It was actually considered for direct use in `thread_rng`, but rejected for performance reasons that shouldn't apply here.

I agree that the use of imports for `OsRng` is preferable. The initial goal was to minimize changes for easier review, but you're right that this is a good time to shore things up.

One additional change that I made at one point was to prefer direct use of `fill_bytes` (provided by `RngCore`) for constructing random byte arrays, instead of relying on `gen` (provided by `Rng`) for a standard distribution. I have no particular reason to distrust the distribution-based approach, but my spidey sense tells me that the more direct and obvious byte-filling approach is a more conservative approach.

**Review by delbonis — DISMISSED** [review]
Concur with suggestions but looks good.

**Comment by storopoli** [comment]
> Concur with suggestions but looks good.

wait for the rebase and the upcoming changes before merging

**Comment by AaronFeickert** [comment]
Marking as draft to address review suggestions and merge conflicts.

**Comment by delbonis** [comment]
I would expect it to be thread safe, it'd be weird if it wasn't.

**Review by delbonis — APPROVED** [review]

**Comment by delbonis** [comment]
@storopoli Are suggestions applied?

**Comment by AaronFeickert** [comment]
> I would expect it to be thread safe, it'd be weird if it wasn't.

If it isn't thread safe, much (most?) of the Rust cryptographic ecosystem is in trouble...

**Review by storopoli — APPROVED** [review]
ACK 8817f4af5677de0a54216994e40b5e5278809143

**Pull request merged**